lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -3771,79 +3642,1099 @@ inline llvm::raw_ostream &operator<<(llvm::raw_ostream &os,
+    emitZeroIndirect(access->getType().getASTType(), access, access->getLoc());
+    diffBuilder.setInsertionPoint(insertionPoint);
+    diffLocalAllocBuilder.createEndAccess(
+        access->getLoc(), access, /*aborted*/ false);
+
+    // Create cleanup for local buffer.
+    differentialLocalAllocations.push_back(newBuf);
+    return (insertion.first->getSecond() = newBuf);
+  }
+
+  //--------------------------------------------------------------------------//
+  // Type transformer
+  //--------------------------------------------------------------------------//
+
+  Optional<VectorSpace> getTangentSpace(CanType type) {
+    return type->getAutoDiffAssociatedTangentSpace(
+        LookUpConformanceInModule(getModule().getSwiftModule()));
+  }
+
+  /// Assuming the given type conforms to `Differentiable` after remapping,
+  /// returns the associated tangent space type.
+  SILType getRemappedTangentType(SILType type) {
+    return SILType::getPrimitiveType(
+        getTangentSpace(remapType(type).getASTType())->getCanonicalType(),
+        type.getCategory());
+  }
+
+  //--------------------------------------------------------------------------//
+  // Managed tangent value mapping
+  //--------------------------------------------------------------------------//
+
+  AdjointValue makeZeroTangentValue(SILType type) {
+    return AdjointValue::createZero(allocator, remapType(type));
+  }
+
+  /// Initializes an original value's corresponding tangent value. Its tangent
+  /// value must not be present before this function is called.
+  void initializeTangentValue(SILBasicBlock *origBB, SILValue originalValue,
+  AdjointValue tangentValue) {
+    assert(origBB->getParent() == original);
+    assert(originalValue->getType().isObject());
+    assert(tangentValue.getType().isObject());
+    assert(originalValue->getFunction() == original);
+    // The tangent value must be in the tangent space.
+    assert(tangentValue.getType() ==
+           getRemappedTangentType(originalValue->getType()));
+    auto insertion =
+        tangentValueMap.try_emplace(originalValue, tangentValue);
+    assert(insertion.second && "Tangent value inserted before");
+  }
+
+  /// Get the tangent for an original value. The given value must be in the
+  /// original function.
+  ///
+  /// This method first tries to find an entry in `tangentValueMap`. If a tangent
+  /// doesn't exist, create a zero tangent.
+  AdjointValue getTangentValue(SILValue originalValue) {
+    assert(originalValue->getType().isObject());
+    assert(originalValue->getFunction() == original);
+    auto insertion = tangentValueMap.try_emplace(
+        originalValue, makeZeroTangentValue(
+        getRemappedTangentType(originalValue->getType())));
+    auto it = insertion.first;
+    return it->getSecond();
+  }
+
+  /// Add a tangent value for the given original value.
+  void addTangentValue(SILBasicBlock *origBB, SILValue originalValue,
```

Comment:
There is no addition for tangent values. Please rename this to `setTangentValue`.